src/scuffed_fuel_estimator.py

```diff
@@ -81,84 +91,91 @@ def prices_to_float(self):
     
     #Getting user inputs for variables.
     def get_car_info(self):
-        self.type = input("What type of fuel do you use? (list below, type the letter)\nSans Plomb 95 E5 (a)\nSans Plomb 95 E10 (b)\nBioethanol E85 (c)\nGazole B7 (d)\nGPL (e)\n").lower()
-        self.refuel = float(input("Last refuel total amount (in L): "))
-        self.fuelamount = float(input("Current fuel amount (in L): "))
-        self.rerefuel = float(input("How much fuel do you plan on for your next refuel ? (in L): "))
+        self.type = input("What type of fuel do you use? (list below, type the letter)\nSans Plomb 95 E5 (a)\nSans Plomb 95 E10 (b)\nBioethanol E85 (c)\nGazole B7 (d)\nGPL (e)\nExit the program (exit)\n").lower()
+        while self.type not in {"a", "b", "c", "d", "e", "exit"}:
+            print("Only values of a, b, c, d or e are allowed!")
+            self.type = input("What type of fuel do you use? (list below, type the letter)\nSans Plomb 95 E5 (a)\nSans Plomb 95 E10 (b)\nBioethanol E85 (c)\nGazole B7 (d)\nGPL (e)\nExit the program (exit)\n").lower()
+        if (self.type == "exit"): 
+            exit("See you soon!")
+        self.refuel = input("Last refuel total amount (in L): ")
+        while not self.refuel.isnumeric():
+            print("Only numeric values are allowed!")
+            self.refuel = input("Last refuel total amount (in L): ")
+        self.refuel = validateFloatInput(self.refuel)
+
+        self.fuelamount = input("Current fuel amount (in L): ")
+        while not self.fuelamount.isnumeric():
+            print("Only numeric values are allowed!")
+            self.fuelamount = input("Current fuel amount (in L): ")
+        self.fuelamount = validateFloatInput(self.fuelamount)
+
+        self.rerefuel = input("How much fuel do you plan on for your next refuel ? (in L): ")
+        while not self.rerefuel.isnumeric():
+            print("Only numeric values are allowed!")
+            self.rerefuel = input("How much fuel do you plan on for your next refuel ? (in L): ")
+        self.rerefuel = validateFloatInput(self.rerefuel)
+
+        print(self.moneyspent())
+        print(self.equivalentmoneyfuel())
+        print(self.equivalentmoneyrefuel())
+        self.__init__()

```

Comment:
@OtterWhisperer Hello and thanks again for your help, I learn a lot of simple basic things that can be useful for every program.
So I understand most of the changes except lines 118-120 : if you call the functions before I'm writing them, how can this work ? I'm trying to understand but I don't get it.

Reply:
Machines can only read zeros and ones. So during the compilation proccess we make sure that our code is compiled into that binary format. Without going into too much detail for example python is an interpreter-based programming language.

Since we compile a file before executing it, we can compare our scenario with the construction of a car, which we compare as our class. The car has certain attributes (class- / instance-variables) and functionalities (methods). These are all constructions which we declare as we write our class. In the compilation/interpretation step the engine makes sure the syntax is correct. As we create an instance of the object car, this instance holds all the methods within, so that we can call them whenever we want. In our case we can call these class-methods during the init-step. Hope it could help you out on that. If you have further questions feel free to ask me on them.

Reply:
Let me give you a better example:

Construction plan of a car --> factory --> objects

1. Construction plan of a car
`class Car {`
`    color = ""`
`}`
We create a class named __Car__ which contains the attribute __color__.

2. Factory or Constructor
What i mean by that is as follows:
Whenever you decide to create an object of a class like `car01 = new Car()` it will call the constructor of that given class (in this case the default constructor).
Constructors in python are defined by the name \_\_init__ and the default constructor is:
`def __init__(self):`. Notice that if you can have multiple constructors with different parameteriation like `def __init__(self, color):`
`.... self.color = color` for example. (pleaes ignore the ....)

3. Objects
We create objects by simply call the new operator `car01 = new Car()`. We save the instance in the variable car01 in this example. Now we can call methods or read/write variables in that object `car01.color = "blue"` and now read the value `print(car01.color)`.

In case you want to colorize your car at the very beginning to specialized needs we can make use of the customized constructors (see the example in step 2). To create an object of a blue car we could write `blueCar = new Car("blue")` for example. Same goes for methods.